projects/portal/src/app/models/cosmos/tx-common.service.ts

```diff
@@ -1,32 +1,186 @@
+              mode: proto.cosmos.tx.signing.v1beta1.SignMode.SIGN_MODE_DIRECT,
+            },
+          },
+          sequence: baseAccount.sequence,
+        },
+      ],
+      fee: {
+        amount: fee ? [fee] : null,
+        gas_limit: cosmosclient.Long.fromString(gas?.amount ? gas.amount : '200000'),
+      },
+    });
+    const txBuilder = new cosmosclient.TxBuilder(sdk, txBody, authInfo);
+    return txBuilder;
+  }
+
+  async signTx(
+    txBuilder: cosmosclient.TxBuilder,
+    signerBaseAccount: proto.cosmos.auth.v1beta1.BaseAccount,
+    currentCosmosWallet: CosmosWallet,
+    privateKey?: string,
+  ): Promise<cosmosclient.TxBuilder | undefined> {
+    if (currentCosmosWallet.type === WalletType.ununifi) {
+      return await this.signTxWithPrivateKey(txBuilder, signerBaseAccount, privateKey);
+    }
+    if (currentCosmosWallet.type === WalletType.keplr) {
+      return this.keplrAppService.signDirect(txBuilder, signerBaseAccount);
+    }
+    if (currentCosmosWallet.type === WalletType.ledger) {
+      return this.signTxWithLedger(txBuilder, signerBaseAccount);
+    }
+    if (currentCosmosWallet.type === WalletType.keyStation) {
+      return this.signTxWithKeyStation(txBuilder, signerBaseAccount);
+    }
+    throw Error('Unsupported wallet type!');
+  }
+
+  async signTxWithPrivateKey(
+    txBuilder: cosmosclient.TxBuilder,
+    signerBaseAccount: proto.cosmos.auth.v1beta1.BaseAccount,
+    privateKey?: string,
+  ): Promise<cosmosclient.TxBuilder | undefined> {
+    let cosmosPrivateKey: cosmosclient.PrivKey | undefined;
+    if (privateKey) {
+      cosmosPrivateKey = createCosmosPrivateKeyFromString(KeyType.secp256k1, privateKey);
+    } else {
+      const privateWallet: StoredWallet & { privateKey: string } =
+        await this.walletAppService.openUnunifiKeyFormDialog();
+      if (!privateWallet || !privateWallet.privateKey) {
+        this.snackBar.open('Failed to get Wallet info from dialog! Tray again!', 'Close');
+        return;
+      }
+      if (!validatePrivateStoredWallet(privateWallet)) {
+        this.snackBar.open('Invalid Wallet info!', 'Close');
+        return;
+      }
+      cosmosPrivateKey = createCosmosPrivateKeyFromString(
+        KeyType.secp256k1,
+        privateWallet.privateKey,
+      );
+      if (!privateKey) {
+        this.snackBar.open('Invalid PrivateKey!', 'Close');
+        return;
+      }
+    }
+    if (!cosmosPrivateKey) {
+      throw Error('Invalid Private Key!');
+    }
+    const signDocBytes = txBuilder.signDocBytes(signerBaseAccount.account_number);
+    const signature = cosmosPrivateKey.sign(signDocBytes);
+    txBuilder.addSignature(signature);
+    return txBuilder;
+  }
+
+  // Todo: This is dummy function and need to implement later.
+  signTxWithKeplr(
+    txBuilder: cosmosclient.TxBuilder,
+    signerBaseAccount: proto.cosmos.auth.v1beta1.BaseAccount,
+  ): cosmosclient.TxBuilder {
+    throw Error('Currently signing with Keplr is not supported!');
+  }
```

Comment:
I guess it is better to remove this function (or use this function to execute keplrAppService.signDirect in this function.)

```suggestion
```